### PR TITLE
Develop backend

### PR DESCRIPTION
**Pull request by Anastasia-Tre**

### CODE REVIEW

File: backend/decode.js [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+
+  if (scodeLen < urlCodeLen) {
+    throw new Error('The code is wrong');
+  }
+
+  const codeOfLink = scode.slice(0, urlCodeLen);
+  const readoutСrc = scode.slice(urlCodeLen);
+
+  const readoutСrcWidth = 2 ** readoutСrc.length;
+  const asciChars = getChunksOfString(codeOfLink, 2);
```
Comment (tedi4t) [inline backend/decode.js]:
```suggestion
  const asciiItemLen = 2; 
  const asciChars = getChunksOfString(codeOfLink, asciiItemLen);
```

File: backend/crc.js [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    for (let b of bytes) {
+      const byte = parseInt(b, 16);
+      crc ^= (byte << (width - 8));
+      const pos = (crc >> (width - 8));
+      crc = ((crc << 8) ^ table[pos]) & castMask;
+    }
+    return (crc ^ finalXorVal);
+  }
+}
+
```
Comment (MaksGovor) [inline backend/crc.js]:
```suggestion
const getCrc = partial(getCrcByConfig, crcParamsByWidth);

```

File: backend/crc.js [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+      castMask = 0xFFFFFFFF;
+      polynom = 0x04C11DB7;
+      initialVal = 0xFFFFFFFF;
+      finalXorVal = 0xFFFFFFFF;
+      break;
+    default:
+      throw "Invalid CRC width";
+      break;
+  }
+  this.table = createCrcTable(width, castMask, polynom);
```
Comment (MaksGovor) [inline backend/crc.js]:
```suggestion
const getCrcByConfig = (crcConfig, width) => {
  let castMask, polynom, initialVal, finalXorVal;
  const paramsByWidth = crcConfig[width];
  if (paramsByWidth) {
    [ castMask, polynom, initialVal, finalXorVal ] = paramsByWidth;
  } else {
    throw 'Invalid CRC width';
  }
  
  const table = createCrcTable(width, castMask, polynom);
  
```

File: backend/crc.js [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+        byte ^= polynom;
+      } else {
+        byte <<= 1;
+      }
+    }
+    crcTable[i] = (byte & castMask);
+  }
+  return crcTable;
+}
+
```
Comment (MaksGovor) [inline backend/crc.js]:
```suggestion

const crcParamsByWidth = {
  '8': [0xFF, 0x07, 0x00, 0x00],
  '16': [0xFFFF, 0x8005, 0x0000, 0x0000],
  '32': [0xFFFFFFFF, 0x04C11DB7, 0xFFFFFFFF, 0xFFFFFFFF]
}

const partial = (fn, ...args) => (...rest) => fn(...args, ...rest);

```

File: backend/decode.js [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+  const generatedCrc = getCrc(readoutСrcWidth)(asciChars);
```
Comment (tedi4t) [inline backend/decode.js]: Maybe we should pass function for crc as a param?

File: backend/decode.js (hunk 1 above)
Comment (MaksGovor) [reply backend/decode.js]: Thank you, accepted.

File: backend/crc.js (hunk 3 above)
Comment (Anastasia-Tre) [reply backend/crc.js]: thanks for review

File: backend/decode.js [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+  if (scodeHex.slice(len - 2) !== END_CODE) {
+    throw new Error('The code reading was not correct');
+  }
+
+  scodeHex = scodeHex.slice(0, len - 2);
+  const res = decodeHexInQueryParam(scodeHex, urlCodeLen);
+  return res;
+};
+
+function getLink(rays, scodeLen) {
```
Comment (tedi4t) [inline backend/decode.js]:
```suggestion
function getLink(rays, linkLen) {
  const asciiItemLen = 2; 
  const scodeLen = linkLen * asciiItemLen;
```

File: backend/decode.js [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+  const stabilized = stabilize(rays);
+
+  const raysCoded = removeGuides(stabilized);
+  const result = decodeDataFromImage(raysCoded, scodeLen * 2);
```
Comment (tedi4t) [inline backend/decode.js]:
```suggestion
  const result = decodeDataFromImage(raysCoded, scodeLen);
```

File: backend/decode.js (hunk 5 above)
Comment (MaksGovor) [reply backend/decode.js]: Yes, I think that would be much better and we would make the code more flexible.

**Review by tedi4t — CHANGES_REQUESTED** [review]

**Review by tedi4t — APPROVED** [review]

File: backend/src/decode.js [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+    if (neededLen.top === top &&
+      neededLen.right === right &&
+      neededLen.bottom === bottom &&
+      neededLen.left === left
+    ) {
+      startFound = true;
+    } else {
+      iterator++;
+    }
+  }
```
Comment (YAGoOaR) [inline backend/src/decode.js]:
```suggestion
  const sides = 4;
  const step = len / sides;

  let startFound = false;
  let iterator = 0;
  const neededLen = [15, 15, 0, 15];

  while (!startFound && iterator < len) {
    const correct = true;

    for(let i = 0; i < sides; i++){
      const ind = (iterator + (step * i)) % len;
      if(neededLen[i] !== rays[ind]){
        correct = false
      }
    }
    if(correct){
      startFound = true;
    }
    else {
      iterator++;
    }
  }
```

File: backend/src/crc.js [hunk 9]
```diff
@@ -0,0 +1,75 @@
+'use strict';
+
+
+class CRC {
+  constructor(width, polynom, initialVal, finalXorVal) {
```
Comment (Anastasia-Tre) [inline backend/src/crc.js]:
```suggestion
  constructor(width, ...params) {
```

File: backend/src/crc.js [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+    if (paramsByWidth) {
+      [this.castMask, this.polynom, this.initialVal, this.finalXorVal] = paramsByWidth;
+    } else {
+      throw 'Invalid CRC width';
+    }
+
+    if (arguments.length === 4) {
+      this.polynom = arguments[1];
+      this.initialVal = arguments[2];
+      this.finalXorVal = arguments[3];
```
Comment (Anastasia-Tre) [inline backend/src/crc.js]:
```suggestion
      [this.polynom,
        this.initialVal,
        this.finalXorVal] = params;
```

**Pull request merged**